File: quote/Projects/PluginModel/ActiveDetail.vb
Imports System.ComponentModel

Imports Model
Imports Model.Common

Public Class ActiveDetail
    Implements INotifyPropertyChanged

    Private _Detail As Detail
    
    private Shared ReadOnly _ActiveDetail As ActiveDetail = New ActiveDetail
    
    Public Shared ReadOnly Property ActiveDetail
        Get 
            Return _ActiveDetail
        End Get
    End Property

    Public Event PropertyChanged As PropertyChangedEventHandler _
        Implements INotifyPropertyChanged.PropertyChanged

    Public Property Detail As Detail
        Get
            Return _Detail
        End Get
        Set(ByVal value As Detail)
            If (_Detail IsNot value) Then
                _Detail = value
                RaiseEvent PropertyChanged(Me, _
                    New PropertyChangedEventArgs("Detail"))
            End If
        End Set
    End Property

End Class

File: quote/Projects/PluginModel/ActiveDetail.cs
namespace Model
{
    using System;
    using System.ComponentModel;
    using System.Linq;
    
    using Model.Common;

    /// <summary>
    /// Holds the <c>detail</c> that is currently active. 
    /// This is a singleton class.
    /// </summary>
    public class ActiveDetail : INotifyPropertyChanged
    {
        /// <summary>
        /// The only instance.
        /// </summary>
        private static ActiveDetail instance = new ActiveDetail();

        /// <summary>
        /// The active detail.
        /// </summary>
        private Detail detail;

        /// <summary>
        /// Event that signals the detail has changed.
        /// </summary>
        public event PropertyChangedEventHandler PropertyChanged;

        /// <summary>
        /// Gets the singleton instance.
        /// </summary>
        /// <value>The instance.</value>
        public static ActiveDetail Instance 
        {
            get { return instance; }
        }

        /// <summary>
        /// Gets or sets the active <c>detail</c>.
        /// </summary>
        /// <value>The detail.</value>
        public Detail Detail 
        {
            get 
            { 
                return this.detail; 
            }
            
            set 
            {
                if (!object.ReferenceEquals(this.detail, value))
                {
                    this.detail = value;
                    if (this.PropertyChanged != null) 
                    {
                        this.PropertyChanged(this, new PropertyChangedEventArgs("Detail"));
                    }
                }
            }
        }
    }
}
